### PR TITLE
Jmalone/bod finetune

### PR DESCRIPTION
**Pull request by jmalone-tt**
### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:jmalone/bod-finetune)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:jmalone/bod-finetune)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:jmalone/bod-finetune)
- [ ] New/Existing tests provide coverage for changes
- [ ] (Optional) Ran [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml) on the PR branch to catch linting errors early

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:jmalone/bod-finetune)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:jmalone/bod-finetune)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:jmalone/bod-finetune)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: tt-train/sources/ttml/models/gpt2.cpp [hunk 1]
```diff
@@ -61,6 +61,8 @@ Transformer::Transformer(const TransformerConfig &config) {
     fmt::print("    Composite layernorm: {}\n", use_composite_layernorm);
     fmt::print("    Weight tying: {}\n", config.weight_tying == WeightTyingType::Enabled ? "Enabled" : "Disabled");
 
+    fmt::print("Rebuild does update");
```
Comment (jmalone-tt) [inline tt-train/sources/ttml/models/gpt2.cpp]: You can delete this line - I was just needing to make sure that rebuilding actually updated. Unfortunately even with an editable pip install, I still needed to rerun `uv pip install --no-build-isolation -e .` to get tt-train updates. I think that's why things weren't being fixed when we ran last time 😱

File: tt-train/configs/training_configs/training_gsm8k_tinyllama.yaml
Comment (ayerofieiev-tt) [inline tt-train/configs/training_configs/training_gsm8k_tinyllama.yaml:1]: file is still called tinyllama btw

File: tt-train/sources/examples/gsm8k_finetune/gsm8k_finetune.py [hunk 2]
```diff
@@ -589,18 +587,19 @@ def train():
...
-                val_batch_generator,
-                testing_data,
-                loss_fn,
-                causal_mask,
-                logits_mask_tensor,
-                max_sequence_length,
-                total_steps,
-            )
-            val_losses.append(last_val_loss)
+            # last_val_loss = validate(
```
Comment (ayerofieiev-tt) [inline tt-train/sources/examples/gsm8k_finetune/gsm8k_finetune.py]: is that expected?

File: tt-train/sources/examples/gsm8k_finetune/gsm8k_finetune.py [hunk 3]
```diff
@@ -480,7 +478,7 @@ def train():
     causal_mask = build_causal_mask(max_sequence_length)
 
     causal_mask = ttml.autograd.Tensor.from_numpy(
```
Comment (philei-tt) [inline tt-train/sources/examples/gsm8k_finetune/gsm8k_finetune.py:495]: You can set casual mask to None, Slava's sdpa will generate it on the fly, this is a lot faster, since this is a requirement for balanced parallelism

File: tt-train/sources/examples/gsm8k_finetune/gsm8k_finetune.py (hunk 2 above)
Comment (jmalone-tt) [reply tt-train/sources/examples/gsm8k_finetune/gsm8k_finetune.py]: This was commented out so we never run validation - significant speedup for demo purposes, but we probably don't want to merge this change
